pkg/operator/machinedeletionhooks/machinedeletionhooks.go

```diff
@@ -0,0 +1,219 @@
+package machinedeletionhooks
+
+import (
+	"context"
+	kerrors "k8s.io/apimachinery/pkg/util/errors"
+	"time"
+
+	machinev1beta1 "github.com/openshift/api/machine/v1beta1"
+	machinev1beta1client "github.com/openshift/client-go/machine/clientset/versioned/typed/machine/v1beta1"
+	machinelistersv1beta1 "github.com/openshift/client-go/machine/listers/machine/v1beta1"
+	"github.com/openshift/cluster-etcd-operator/pkg/etcdcli"
+	"github.com/openshift/cluster-etcd-operator/pkg/operator/ceohelpers"
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	operatorv1helpers "github.com/openshift/library-go/pkg/operator/v1helpers"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/sets"
+	corev1listers "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+)
+
+type machineDeletionHooksController struct {
+	etcdClient                            etcdcli.EtcdClient
+	machineClient                         machinev1beta1client.MachineInterface
+	masterMachineLister                   machinelistersv1beta1.MachineLister
+	configMapListerForKubeSystemNamespace corev1listers.ConfigMapNamespaceLister
+	// machineAPIChecker determines if the precondition for this controller is met,
+	// this controller can be run only on a cluster that exposes a functional Machine API
+	machineAPIChecker     ceohelpers.MachineAPIChecker
+	masterMachineSelector labels.Selector
+}
+
+// NewMachineDeletionHooksController reconciles machine hooks for master machines
+//
+// note:
+//  The Machine Deletion Hook this controller reconciles is a mechanism within the Machine API that allow this operator to hold up removal of a machine
+//  until a replacement member has been promoted to a voting member.
+func NewMachineDeletionHooksController(
+	operatorClient operatorv1helpers.OperatorClient,
+	machineClient machinev1beta1client.MachineInterface,
+	etcdClient etcdcli.EtcdClient,
+	machineAPIChecker ceohelpers.MachineAPIChecker,
+	masterMachineSelector labels.Selector,
+	kubeInformersForNamespaces operatorv1helpers.KubeInformersForNamespaces,
+	masterMachineInformer cache.SharedIndexInformer,
+	eventRecorder events.Recorder) factory.Controller {
+	c := &machineDeletionHooksController{
+		machineClient:                         machineClient,
+		etcdClient:                            etcdClient,
+		machineAPIChecker:                     machineAPIChecker,
+		masterMachineSelector:                 masterMachineSelector,
+		masterMachineLister:                   machinelistersv1beta1.NewMachineLister(masterMachineInformer.GetIndexer()),
+		configMapListerForKubeSystemNamespace: kubeInformersForNamespaces.InformersFor("kube-system").Core().V1().ConfigMaps().Lister().ConfigMaps("kube-system"),
+	}
+	return factory.New().
+		WithSync(c.sync).
+		WithSyncDegradedOnError(operatorClient).
+		WithInformers(
```

Comment:
bareinformers for the machinelister?

Reply:
why? we do want to react to changes in `masterMachineInformer`

Reply:
I would expect this operator to react to events from master machines so that if a user deletes the hook, it can re-add it as per the requirements (ie re-add it when the machine is not yet deleted)